---
title: 'CAV Cohort 2 - RNAseq - Whole Blood - QC'
author: "Samuel O Skinner"
date: "March 3, 2017"
output:
  html_document:
    number_sections: yes
---


```{r global_opts, echo=FALSE, cache=FALSE}
knitr::opts_chunk$set(fig.width=8, fig.height=4, fig.align='center',
                      echo=TRUE, warning=FALSE, message=FALSE,
                      cache=FALSE, autodep=TRUE)
# knitr::opts_knit$set(root.dir = "..")

## numbers >= 10^5 will be denoted in scientific notation,
## and rounded to 2 digits
options(scipen = 1, digits = 5)
```


# Summary
This dataset contains the Whole Blood samples from Cohort 2 from the Core for Assay Validation (CAV) project.

----- 
\  
# R environment set-up
\  
## Loading packages

```{r load_packages, cache=FALSE}
# Clean up the environment
rm(list = ls())
cleanup <- gc(verbose = FALSE)

# Load libraries I'll need here
library(goseq)
library(GO.db)
library(annotables)
library(dplyr)
library(limma)
library(edgeR)
library(readr)
library(readxl)
library(ggplot2)
library(cowplot)
library(reshape2)

# Packages for R markdown stuff
library(knitr)
library(shiny)
```


\  

## Loading data

Load the counts and metrics data for the project into R, along with sample annotation for project libraries. Highly variable genes from Cohort 0 (P108) will be filtered out in this analysis for consistency.

```{r load_data}
# Read CSV file of flow cell with read counts
countFile <- "data/MultiFlowcellProjects_combined_counts_merged.csv"
countDat <- read_csv(countFile) # 64253 obs. of  499 variables
#dim(countDat)

# Read CSV file of flow cell with RNAseq/alignment metrics
metricFile <- "data/MultiFlowcellProjects_combined_metrics_merged.csv"
metricDat <- read_csv(metricFile) # 498 obs. of  71 variables
#dim(metricDat)


# Read RData file of flow cell with Sample name
load("../CAV-Cohort-2-RNAseq-CellSubsets/data/Participant_Data_TrialShare/Lim_data_trimmed_2015-12-31.RData")
master <- master[!is.na(master$libID),]
master[master$Study=="AbATE",]$Study <- "ABATE"
master <- master[master$Study %in% c("START","T1DAL","ABATE"),]
designDat <- master[master$Weeks==0,]
designDat <- designDat %>% select(-c(`2hrAUC`, Baseline, Diff, Percent, Responder))


# Load in T1DAL unmasking csv and add participant_ids to designDat
unmaskFile <- "data/T1DAL_unmask.csv"
unmaskDat  <- read_csv(unmaskFile)
unmaskDat <- unmaskDat[order(unmaskDat$Masked_ID),]
unmaskDat <- unmaskDat[unmaskDat$Masked_ID %in% designDat$PatientID,]

all(designDat[designDat$Study=="T1DAL",]$PatientID == unmaskDat$Masked_ID)

designDat[designDat$Study=="T1DAL",]$PatientID <- unmaskDat$Participantid

designDat <- designDat[order(designDat$PatientID),]
designDat <- designDat[order(designDat$Study),]


# Add in the investigator ID to the T1DAL patientID's
demoDat <- read_excel("../CAV-Cohort-2-DemographicData/data/20170110 ITN clinical and demographic.xlsx")
demoDat <- demoDat %>% select(`Participant ID`, ID, `Investigator ID`)
demoDat$Study <- unlist(lapply(demoDat$`Participant ID`, function(x) {strsplit(x, "_")[[1]][[1]]}))
demoDat <- demoDat[!duplicated(demoDat$`Participant ID`),]

demoDat <- demoDat[order(demoDat$`Investigator ID`),]
demoDat <- demoDat[order(demoDat$Study),]

# Drop some of the libraries for people not in the demo file
designDat <- designDat[-which(designDat$Study=="T1DAL" & !(designDat$PatientID %in% demoDat$`Investigator ID`)),]

# Finish up the participant_ids
designDat$participant_id <- apply(designDat, 1, function(x) {paste(x[1],x[4],sep="_")})
i_T <- which(designDat$Study=="T1DAL")
designDat[i_T,]$participant_id <- demoDat[demoDat$Study=="T1DAL",]$`Participant ID`

# Go back to sorting by library
designDat <- designDat[order(designDat$libID),]


# Read CSV file with highly variable genes from Cohort 0
genebcvFile <- "../CAV-Cohort-0-RNAseq-WholeBlood/data/CAV-Cohort-0-RNAseq-WholeBlood-HighCVAnalytes_20170124b.csv"
genebcvDat <- read_csv(genebcvFile) # 2704 obs. of 7 variables
#dim(genebcvDat)

```


\


## Data munging

Start by reformatting some of the variable and columns names for unity and clarity. Move to structuring some kind of experimental design data structure.

```{r clean_data, echo=FALSE}

# Scrape the gene names off one countDat and into geneDat, check to make sure the rows are the same
geneDat <- data_frame(ensgene = countDat$geneName)
countDat <- countDat %>% dplyr::select(-geneName)

# Reformat variable names in metricDat
names(metricDat) <- names(metricDat) %>%
  tolower() %>%
  make.unique(sep = "_") # de-dup variable names

# Rename library variable in countDat to only have libXXXX 
lib_names <- colnames(countDat)
first_names <- sapply(strsplit(lib_names, "_"), function (x) {x[1]})
colnames(countDat) <- first_names

# Rename library variable in metricDat to only have libXXXX 
lib_names <- metricDat$libid
first_names <- sapply(strsplit(lib_names, "_"), function (x) {x[1]})
metricDat$libid <- first_names

# Replace special chars and spaces with underscores and lowercase names in demoDat
names(designDat) <- names(designDat) %>% tolower()

# Order designDat by libid and remove rows with NA for libid
designDat <- designDat[order(designDat$libid),]

# Reduce countDat down to the libid's in designDat and order
countDat <- countDat[,colnames(countDat) %in% designDat$libid]
countDat <- countDat[,order(colnames(countDat))]

# Reduce metricDat down to the libid's in designDat and order
metricDat <- metricDat[metricDat$libid %in% designDat$libid,]
metricDat <- metricDat[order(metricDat$libid),]

# Check to see if all of the libraries are represented
all(designDat$libid == colnames(countDat))
all(designDat$libid == metricDat$libid)


```


\  

# General QC

```{r inspect_data}

# Make summary data structure
metricSummary <- metricDat %>%
  mutate(percent_duplication = unpaired_read_duplicates / unpaired_reads_examined) %>%
  mutate(percent_aligned_w_dups = unpaired_reads_examined / fastq_total_reads) %>% 
  select(libid, median_cv_coverage, fastq_total_reads, percent_aligned_w_dups, percent_duplication)

```
\  

## Inspection of various quality metrics and distributions

\  

### Inspect the number of total FASTQ reads
```{r plot_metric_1, echo=FALSE}

barplot(sort(metricSummary$fastq_total_reads)/1E6,
        xlab="Libraries",
        ylab="Total FASTQ Reads (millions)",
        main = "Total FASTQ Reads")
abline(h=1)

```

\  

### Inspect library median CV and percent aligned
```{r plot_metric_2, echo=FALSE}

plot(x=metricSummary$median_cv_coverage, 
     y=metricSummary$percent_aligned_w_dups, 
     xlab="Median CV coverage", 
     ylab="Percent Aligned",
     xlim=c(0,1),
     ylim=c(0.8,1),
     main="Percent Aligned vs. Median CV Coverage")
abline(h=0.8)
abline(v=1.0)
#rect(0,0.8,1,1,col=rgb(0.5,0.5,0.5,1/4), border=NA)

```





\  


## Gene filtering and inspection of resulting count and cpm distributions

\  

### Setup dge structure and annotate genes
```{r gene_annotation}

# Set up annotation list
ens2Gene <- grch38 %>% select(ensgene, entrez, symbol, biotype, chr)

# Remove duplicates from multiple entrez entries per ensgene entries
ens2Gene <- ens2Gene[!duplicated(ens2Gene$ensgene),]

# Join with geneDat
geneDat <- geneDat %>% merge(ens2Gene, by="ensgene", all.x = TRUE)

# Set up DGEList
dge <- DGEList(counts = countDat, genes = geneDat)


```

\  

### Filter genes
``` {r gene_filtering}

#Current size
dim(dge)

# Keep all genes that have at least 1 count per million in at least 20% of libraries
keepRows <- rowSums(cpm(dge$counts) >= 1) >= 0.2*ncol(dge)
curDGE <- dge[keepRows, ]
dim(curDGE)

# Discard all genes that are highly variable
discardrows <- which(curDGE$genes$ensgene %in% genebcvDat$ensgene)
curDGE <- curDGE[-discardrows,]
dim(curDGE)

# Use only the protein-coding genes
i_pc  <- which(curDGE$genes$biotype == "protein_coding")
curDGE <- curDGE[i_pc,]
dim(curDGE)

# reset library sizes
curDGE$samples$lib.size <- colSums(curDGE$counts)

# calculate normalization factors (effective library size = lib.size * norm.factor)
curDGE <- calcNormFactors(curDGE)

# Voom transformation
curDGE <- voom(curDGE)

```




# Basic analysis

\  

## Principal component analysis (PCA)
Performing the PCA calculation on log counts. 
```{r initial_pca}

# Compute principal components
pca_log <- prcomp(t(curDGE$E))

pca_log$x <- as.data.frame(pca_log$x)


```


The relative variance explained by each PC.

``` {r display_PC_std}
percVariance <- pca_log$sdev^2 / sum(pca_log$sdev^2)

pc_df <- data_frame(pc = 1:length(pca_log$sdev), 
                    percVariance)

ggplot(pc_df, aes(x=pc, y=percVariance)) + 
  geom_bar(stat="identity")


```





The first four PC's against each other.
```{r plot_initial_pca, echo=FALSE}



Study <- designDat$study


plot1.2 <- ggplot(pca_log$x, 
                  aes(x=PC1,
                      y=PC2,
                      colour=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))

plot1.3 <- ggplot(pca_log$x, 
                  aes(x=PC1,
                      y=PC3,
                      color=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))

plot1.4 <- ggplot(pca_log$x, 
                  aes(x=PC1,
                      y=PC4,
                      color=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))

plot2.3 <- ggplot(pca_log$x, 
                  aes(x=PC2,
                      y=PC3,
                      color=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))

plot2.4 <- ggplot(pca_log$x, 
                  aes(x=PC2,
                      y=PC4,
                      color=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))

plot3.4 <- ggplot(pca_log$x, 
                  aes(x=PC3,
                      y=PC4,
                      color=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))



grobs <- ggplotGrob(plot1.2 + theme(legend.position="top"))$grobs
legend_b <- grobs[[which(sapply(grobs, function(x) x$name) == "guide-box")]]

prow <- plot_grid(plot1.2, plot1.3, plot1.4, 
                  plot2.3, plot2.4, plot3.4,
                  ncol=3)
plot_grid(legend_b, prow, ncol=1, rel_heights=c(.1, 1))


```

Probably will want to batch correct for study before shipping it off to Dror.


## Batch correct for study

```{r batch correct}

curDGE$E <- removeBatchEffect(curDGE$E, batch=designDat$study)

```


```{r re-display pca, echo=FALSE}


# Compute principal components
pca_log <- prcomp(t(curDGE$E))

pca_log$x <- as.data.frame(pca_log$x)

Study <- designDat$study


plot1.2 <- ggplot(pca_log$x, 
                  aes(x=PC1,
                      y=PC2,
                      colour=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))

plot1.3 <- ggplot(pca_log$x, 
                  aes(x=PC1,
                      y=PC3,
                      color=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))

plot1.4 <- ggplot(pca_log$x, 
                  aes(x=PC1,
                      y=PC4,
                      color=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))

plot2.3 <- ggplot(pca_log$x, 
                  aes(x=PC2,
                      y=PC3,
                      color=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))

plot2.4 <- ggplot(pca_log$x, 
                  aes(x=PC2,
                      y=PC4,
                      color=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))

plot3.4 <- ggplot(pca_log$x, 
                  aes(x=PC3,
                      y=PC4,
                      color=Study)) + 
  geom_point() +
  theme(legend.position="none",
        text = element_text(size=10),
        axis.text = element_text(size=8))



grobs <- ggplotGrob(plot1.2 + theme(legend.position="top"))$grobs
legend_b <- grobs[[which(sapply(grobs, function(x) x$name) == "guide-box")]]

prow <- plot_grid(plot1.2, plot1.3, plot1.4, 
                  plot2.3, plot2.4, plot3.4,
                  ncol=3)
plot_grid(legend_b, prow, ncol=1, rel_heights=c(.1, 1))

```



# Write output to file
```{r write_output_file}

colnames(curDGE$E) <- designDat$participant_id


save(curDGE, file = "data/CAV-Cohort-2-RNAseq-WholeBlood_data_20170516b.RData")

```

